flocker/node/agents/blockdevice.py

```diff
@@ -0,0 +1,270 @@
+
+    :param unicode blockdevice_id: The unique identifier of the block device.
+    """
+    def __init__(self, blockdevice_id):
+        if not isinstance(blockdevice_id, unicode):
+            raise TypeError(
+                'Unexpected blockdevice_id type. '
+                'Expected unicode. '
+                'Got {!r}.'.format(blockdevice_id)
+            )
+        Exception.__init__(self, blockdevice_id)
+        self.blockdevice_id = blockdevice_id
+
+
+class UnknownVolume(VolumeException):
+    """
+    The block device could not be found.
+    """
+
+
+class AlreadyAttachedVolume(VolumeException):
+    """
+    A failed attempt to attach a block device that is already attached.
+    """
+
+
+class UnattachedVolume(VolumeException):
+    """
+    An attempt was made to operate on an unattached volume but the operation
+    requires the volume to be attached.
+    """
+
+
+class IBlockDeviceAPI(Interface):
+    """
+    Common operations provided by all block device backends.
+    """
+    def create_volume(size):
+        """
+        Create a new block device.
+
+        :param int size: The size of the new block device in bytes.
+        :returns: A ``BlockDeviceVolume``.
+        """
+
+    def attach_volume(blockdevice_id, host):
+        """
+        Attach ``blockdevice_id`` to ``host``.
+
+        :param unicode blockdevice_id: The unique identifier for the block
+            device being attached.
+        :raises UnknownVolume: If the supplied ``blockdevice_id`` does not
+            exist.
+        :raises AlreadyAttachedVolume: If the supplied ``blockdevice_id`` is
+            already attached.
+        :returns: A ``BlockDeviceVolume`` with a ``host`` attribute set to
+            ``host``.
+        """
+
+    def list_volumes():
+        """
+        List all the block devices available via the back end API.
+
+        :returns: A ``list`` of ``BlockDeviceVolume``s.
+        """
+
+    def get_device_path(blockdevice_id):
+        """
+        Calculate the path at which ``blockdevice_id`` will be exposed on the
+        host when attached.
+
+        :param unicode blockdevice_id: The unique identifier for the block
+            device.
+        :returns: A ``FilePath`` for the device.
+        """
+
+
+class BlockDeviceVolume(PRecord):
+    """
+    A block device that may be attached to a host.
+
+    :ivar unicode blockdevice_id: The unique identifier of the block device.
+    :ivar int size: The size of the block device.
+    :ivar bytes host: The IP address of the host to which the block device is
+        attached or ``None`` if it is currently unattached.
+    """
+    blockdevice_id = field(type=unicode, mandatory=True)
+    size = field(type=int, mandatory=True)
+    # XXX: Should be hostname, for consistency of is host a better name since
+    # we currently only expect IP addresses?
+    host = field(type=(bytes, type(None)), initial=None)
```

Comment:
I tried this, but it makes the tests difficult, because the device_path can't be predicted...and in trying it this way, I've convinced myself that the original get_device_path API is correct.